feat(navigator): add course hold mode (#27156)
PX4/PX4-Autopilot: docs/en/flight_modes_fw/guided_course.md

:::

::: info

- Requires a horizontal velocity estimate (e.g. GPS/dead-reckoning).
  Course commands will be rejected if the velocity estimate is unavailable.
- Manual control input is ignored.

<!-- https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/commander/ModeUtil/mode_requirements.cpp -->

:::

## Overview

On activation, the vehicle captures its current velocity over ground vector as the initial course bearing and holds altitude and airspeed from the moment of activation.
The vehicle then flies that course indefinitely until the operator issues a new command.

There is no waypoint sequencing or autonomous path planning: the GCS guides the vehicle in real time by sending individual commands.

## Supported Commands

The following commands are accepted while in Guided Course mode:

| Command                      | Effect                                                                                                  |
| ---------------------------- | ------------------------------------------------------------------------------------------------------- |

> | `MAV_CMD_GUIDED_CHANGE_HEADING`   | with `HEADING_TYPE = 0`. Set a new course bearing (degrees, 0 = north). Rejected if horizontal velocity estimate is unavailable. |

Reviewer: ```suggestion
| [MAV_CMD_GUIDED_CHANGE_HEADING]   |  Set a new course bearing (degrees, 0 = north) with `HEADING_TYPE = 0`. Rejected if horizontal velocity estimate is unavailable. |
```

| [MAV_CMD_DO_CHANGE_ALTITUDE] | Set a new target altitude (AMSL, metres).                                                               |
| [MAV_CMD_DO_CHANGE_SPEED]    | Set a new equivalent airspeed via param2 (m/s). If param2 ≤ 0, the default cruise speed is restored.    |

> [MAV_CMD_DO_CHANGE_ALTITUDE]: https://mavlink.io/en/messages/common.html#MAV_CMD_DO_CHANGE_ALTITUDE

Reviewer: Assuming it moves as planned.
```suggestion
[MAV_CMD_GUIDED_CHANGE_HEADING]: https://mavlink.io/en/messages/common.html#MAV_CMD_GUIDED_CHANGE_HEADING
[MAV_CMD_DO_CHANGE_ALTITUDE]: https://mavlink.io/en/messages/common.html#MAV_CMD_DO_CHANGE_ALTITUDE
```

:::

::: info

- Requires a horizontal velocity estimate (e.g. GPS/dead-reckoning).
  Course commands will be rejected if the velocity estimate is unavailable.
- Manual control input is ignored.

<!-- https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/commander/ModeUtil/mode_requirements.cpp -->

:::

## Overview

On activation, the vehicle captures its current velocity over ground vector as the initial course bearing and holds altitude and airspeed from the moment of activation.
The vehicle then flies that course indefinitely until the operator issues a new command.

There is no waypoint sequencing or autonomous path planning: the GCS guides the vehicle in real time by sending individual commands.

## Supported Commands

The following commands are accepted while in Guided Course mode:

| Command                      | Effect                                                                                                  |
| ---------------------------- | ------------------------------------------------------------------------------------------------------- |

> | `MAV_CMD_DO_CHANGE_COURSE`   | Set a new course bearing (degrees, 0 = north). Rejected if horizontal velocity estimate is unavailable. |

Reviewer 2: ```suggestion
| `MAV_CMD_GUIDED_CHANGE_HEADING`   | with `HEADING_TYPE = 0`. Set a new course bearing (degrees, 0 = north). Rejected if horizontal velocity estimate is unavailable. |
```

> | [MAV_CMD_GUIDED_CHANGE_HEADING]   |  Set a new course bearing (degrees, 0 = north) with `HEADING_TYPE = 0`. Rejected if horizontal velocity estimate is unavailable. |

Reviewer: Prettier the page before merging please.

| [MAV_CMD_DO_CHANGE_ALTITUDE] | Set a new target altitude (AMSL, metres).                                                               |
| [MAV_CMD_DO_CHANGE_SPEED]    | Set a new equivalent airspeed via param2 (m/s). If param2 ≤ 0, the default cruise speed is restored.    |

[MAV_CMD_DO_CHANGE_ALTITUDE]: https://mavlink.io/en/messages/common.html#MAV_CMD_DO_CHANGE_ALTITUDE

> [MAV_CMD_DO_REPOSITION]: https://mavlink.io/en/messages/common.html#MAV_CMD_DO_REPOSITION

Reviewer 2: ```suggestion
```

# Guided Course Mode (Fixed-Wing)

<img src="../../assets/site/position_fixed.svg" title="Position required (e.g. GPS)" width="30px" />

_Guided Course mode_ maintains a constant ground track (course), altitude, and airspeed without any manual stick input.

> The operator controls the vehicle entirely via GCS commands, making it the guided equivalent of [Position mode](../flight_modes_fw/position.md).

Reviewer: Postion mode holds the ground track - so if you had a strong sideways wind the plane would either hold or return to that actual track. Just confirming same for this?

We should link to the commands allowed

```suggestion
_Guided Course mode_ maintains a constant ground track (course), altitude, and airspeed without any manual stick input.
The operator controls the vehicle entirely via [GCS commands](#in-flight-commands), making it the guided equivalent of [Position mode](../flight_modes_fw/position.md).
```

Reviewer 2: Sorry, missed this comment. 

> Just confirming same for this?

yes :smile:

:::tip
This mode is suited to situations where an operator wants to guide a fixed-wing vehicle from a GCS without manual control.
:::

::: info

- Requires a horizontal velocity estimate (e.g. GPS/dead-reckoning).
  Course commands will be rejected if the velocity estimate is unavailable.

> - RC stick movement is ignored.

Reviewer 3: ```suggestion
- Manual control input is ignored.
```

<!-- https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/commander/ModeUtil/mode_requirements.cpp -->

:::

## Overview

> On activation, the vehicle captures its current velocity vector as the initial course bearing and holds altitude and airspeed from the moment of activation.

Reviewer 3: ```suggestion
On activation, the vehicle captures its current velocity over ground vector as the initial course bearing and holds altitude and airspeed from the moment of activation.
```

The vehicle then flies that course indefinitely until the operator issues a new command.

There is no waypoint sequencing or autonomous path planning: the GCS guides the vehicle in real time by sending individual commands.

> ## In-flight commands

Reviewer: How about (and fix link I suggested above)

```suggestion
## Supported Commands
```

The following commands are accepted while in Guided Course mode:

| Command                      | Effect                                                                                                  |
| ---------------------------- | ------------------------------------------------------------------------------------------------------- |

> | `MAV_CMD_DO_CHANGE_COURSE`   | Set a new course bearing (degrees, 0 = north). Rejected if horizontal velocity estimate is unavailable. |

Reviewer: @mahima-yoga This all looks cool. I did a minor subedit - ran prettier.

Also added these links.  Didn't do this CHANGE_COURSE because it doesn't exist.

Reviewer 2: tada! https://github.com/mavlink/mavlink/pull/2487

Reviewer: It's like magic :-). See https://github.com/mavlink/mavlink/pull/2487/changes#r3232321039 for my due diligence.

| [MAV_CMD_DO_CHANGE_ALTITUDE] | Set a new target altitude (AMSL, metres).                                                               |

> | [MAV_CMD_DO_REPOSITION]      | Set a new target altitude via param7.                                                                   |

Reviewer 3: Do you want to listen to that one additionally to `MAV_CMD_DO_CHANGE_ALTITUDE`?

> | [MAV_CMD_DO_REPOSITION]      | Set a new target altitude via param7.                                                                   |

Reviewer 2: ```suggestion
```
